---
title: Annotating Neutrophil eQTL with Blueprint Data
author: Peter Humburg
date: `r format(Sys.time(), "%a %d %b %Y")`
---

```{r setup, include=FALSE}
library(GenomicRanges)
library(ggplot2)
library(pander)

opts_knit$set(root.dir = "/well/hill/Vivek/eQTL")
opts_knit$set(autodep=TRUE)
opts_knit$set(tidy=TRUE)
```

# Introduction
The aim of this analysis is to annotate eQTL obtained from the analysis
of neutrophils with the results of bisulfite sequencing and ChIP-seq
for several histone modifications generated by the Blueprint consortium.
BED files with ChIP-seq peaks and hyper- and hypo-methylated regions
all neutrophil samples were obtained form the Blueprint ftp server.

All eQTL significant at an FDR threshold of 0.05 will be annotated with
the distance to the closest hyper/hypo methylated site as well as the distance
to the available histone marks. The subsequent analysis of these distances
focuses on eQTL of particular interest, e.g. the lead SNP for each gene or
all peak eQTL.

```{r loadEQTL, echo=FALSE, cache=TRUE}
load("blueprint/Neut_results/Neut_neutonly_15PC_140203.rdata")
snppos <- read.table("Neut/Matrix_eQTL_SNP_loc_hg19_090912.txt", header=TRUE,
		stringsAsFactors=FALSE)
cis <- merge(snppos, subset(me$cis$eqtls, FDR < 0.05), by.x="snp", by.y="snps")
cis <- with(cis, GRanges(seqnames=Rle(paste0("chr", chrm_snp)), 
		ranges=IRanges(start=pos, width=1),
		snp=snp, gene=gene, statistic=statistic, pvalue=pvalue,
		FDR=FDR, beta=beta))
trans <- merge(snppos, subset(me$trans$eqtls, FDR < 0.05), by.x="snp", by.y="snps")
trans <- with(trans, GRanges(seqnames=Rle(paste0("chr", chrm_snp)), 
		ranges=IRanges(start=pos, width=1),
		snp=snp, gene=gene, statistic=statistic, pvalue=pvalue,
		FDR=FDR, beta=beta))

rm(me)
```

# Methylation data
## Data processing
The BED files provided by Blueprint do not conform to the BED specification.
This causes difficulties when trying to process them with BED file parsers.
To remedy this the first five columns of each file are extracted prior
to further analysis.

```{r reformatBED, engine="bash", echo=FALSE, cache=TRUE}
for file in $(ls blueprint/WGBS_Neut/*.bed)
do
	cut -f1-5 $file > ${file}5
done	
```
The re-formatted BED files are imported into R for further processing with 
Bioconductor. 

```{r loadWGBS, echo=FALSE, cache=TRUE}
methFiles <- dir("blueprint/WGBS_Neut", ".bed5", full.names=TRUE)
splitNames <- strsplit(basename(methFiles), ".", fixed=TRUE)
sample <- sapply(splitNames, "[[", 1)
type <- sapply(splitNames, "[[", 2)
methDF <- data.frame(file=methFiles, sample=sample, type=type)
methGR <-lapply(methFiles, rtracklayer::import.bed, genome="hg19", asRangedData=FALSE)
```

## Defining regions of hyper- and hypo-methylation
Regions considered to be either hyper- or hypo-methylated are defined as follows.
A genomic region is considered to be hyper-methylated in neutrophils if it
meets the criteria for hyper-methylation used in the analysis of the Blueprint consortium
(average methylation of >0.75 and all CpGs in the regions have methylation >0.5)
in at least two neutrophil samplesand the region isn't hypo-methylated in any of 
the neutrophil samples. Hypo-methylated regions are defined accordingly.

```{r defineRegions, echo=FALSE, cache=TRUE}
allHypo <- do.call(union, c(methGR[methDF$type == "hypo_meth"], ignore.strand=TRUE))
allHyper <- do.call(union, c(methGR[methDF$type == "hyper_meth"], ignore.strand=TRUE))
hypo <- lapply(methGR[methDF$type == "hypo_meth"], setdiff, allHyper)
hyper <- lapply(methGR[methDF$type == "hyper_meth"], setdiff, allHypo)
```

```{r regionLength, echo=FALSE, fig.cap="Length distribution of hyper- and hypo-methylated regions"}
regLen <- list()
regLen$hypo <- data.frame(length=unlist(lapply(hypo, width)), type="hypo methylated")
regLen$hyper <- data.frame(length=unlist(lapply(hyper, width)), type="hyper methylated")
regLen <- rbind(regLen$hypo, regLen$hyper)
ggplot(regLen, aes(x=length, y=..density..)) + geom_histogram(binwidth=100) + facet_grid( ~ type) + 
		theme_bw() + xlim(0,7e4) + ylim(0,0.00075)
```

Considering the length distribution of hyper- and hypo-methylated regions it is apparent
that a lot of regions are quite short while some, especially for hyper-methylation,
are quite large. This is further emphasised by the following summary table.

```{r lengthSummary, echo=FALSE, results="asis"}
lengthSum <- tapply(regLen$length, regLen$type, summary)
set.caption("Summary of length distribution for hypo- and hyper-methylated regions.")
pander(do.call(cbind, lengthSum), style="rmarkdown")
```

Based on this we set the minimum overlap between regions from different individuals
to 50bp. 
```{r overlapRegions, echo=FALSE, cache=TRUE, fig.caption="Length distribution of consensus regions."}
hypo <- Reduce(function(x,y) subsetByOverlaps(x,y, minoverlap=50L), hypo)
hyper <- Reduce(function(x,y) subsetByOverlaps(x,y, minoverlap=50L), hyper)
regLen.comb <- data.frame(length=c(width(hypo), width(hyper)), 
		type=c(rep("hypo methylated", length(hypo)), rep("hyper methylated", length(hyper))),
		consensus=TRUE)
regLen$consensus <- FALSE
regLen.comb <- rbind(regLen.comb, regLen)
regLen.comb$type <- factor(as.character(regLen.comb$type), 
		levels=c("hypo methylated", "hyper methylated"))
ggplot(regLen.comb, aes(x=length, colour=consensus)) + 
		geom_density() + facet_grid( ~ type) + 
		theme_bw() + xlim(0,7e4) + ylim(0,0.00075)
```

While this excludes small regions the effect on the length of the remaining regions is limited,
consistent with substantial overlaps between samples.

```{r lengthSummaryConsensus, echo=FALSE, results="asis"}
lengthSum.comb <- with(subset(regLen.comb, consensus), tapply(length, type, summary))
set.caption("Summary of length distribution for hypo- and hyper-methylated consensus regions.")
pander(do.call(cbind, lengthSum.comb), style="rmarkdown")
```

## Annotating eQTL
All significant eQTL are annotated with their distance to the closest hyper- and 
hypo-methylated regions.

```{r methDist, echo=FALSE, cache=TRUE}
metaCis <- mcols(cis)
metaCis$hypo_meth <- mcols(distanceToNearest(cis, hypo))$distance
metaCis$hyper_meth <- mcols(distanceToNearest(cis, hyper))$distance
mcols(cis) <- metaCis
metaTrans <- mcols(trans)
metaTrans$hypo_meth <- mcols(distanceToNearest(trans, hypo))$distance
metaTrans$hyper_meth <- mcols(distanceToNearest(trans, hyper))$distance
mcols(trans) <- metaTrans
```

```{r methDistLeadPlot, echo=FALSE, fig.cap="Distance from lead SNP to nearest methylation island,"}
cisLeads <- by(mcols(cis), mcols(cis)$gene, function(x) x$snp[which.min(x$pvalue)])
transLeads <- by(mcols(trans), mcols(trans)$gene, function(x) x$snp[which.min(x$pvalue)])
cisDF <- mcols(subset(cis, snp %in% cisLeads))
cisDF <- data.frame(eQTL="cis", feature=c(rep("hypo methylation", nrow(cisDF)),
			rep("hyper methylation", nrow(cisDF))), 
		distance=c(cisDF$hypo_meth, cisDF$hyper_meth), stringsAsFactors=FALSE)
transDF <- mcols(subset(trans, snp %in% transLeads))
transDF <- data.frame(eQTL="trans", feature=c(rep("hypo methylation", nrow(transDF)),
			rep("hyper methylation", nrow(transDF))), 
		distance=c(transDF$hypo_meth, transDF$hyper_meth), stringsAsFactors=FALSE)
methDistDF <- rbind(cisDF, transDF)
ggplot(methDistDF, aes(x=distance)) + geom_density() + facet_grid(eQTL ~ feature) +
		theme_bw() 
```

# Appendix {-}
## Session Info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```
