---
title: Annotating Neutrophil eQTL with Blueprint Data
author: Peter Humburg
date: `r format(Sys.time(), "%a %d %b %Y")`
---

```{r setup, include=FALSE}
library(GenomicRanges)
library(rtracklayer)

opts_knit$set(root.dir = "/well/hill/Vivek/eQTL")
opts_knit$set(autodep=TRUE)
opts_knit$set(tidy=TRUE)
```

# Introduction
The aim of this analysis is to annotate eQTL obtained from the analysis
of neutrophils with the results of bisulfite sequencing and ChIP-seq
for several histone modifications generated by the Blueprint consortium.
BED files with ChIP-seq peaks and hyper- and hypo-methylated regions
all neutrophil samples were obtained form the Blueprint ftp server.

# Methylation data
## Data processing
The BED files provided by Blueprint do not conform to the BED specification.
This causes difficulties when trying to process them with BED file parsers.
To remedy this the first four columns of each file are extracted prior
to further analysis.

```{r reformatBED, engine="bash", echo=FALSE}
for file in $(ls blueprint/WGBS_Neut/*.bed)
do
	cut -f1-4 $file > ${file}4
done	
```

```{r loadWGBS, echo=FALSE}
methFiles <- dir("blueprint/WGBS_Neut", ".bed4", full.names=TRUE)
splitNames <- strsplit(basename(methFiles), ".")
sample <- sapply(splitNames, "[[", 1)
type <- sapply(splitNames, "[[", 2)
methDF <- data.frame(file=methFiles, sample=sample, type=type)
methGR <- lapply(methFiles, import.bed, genome="hg19", asRangedData=FALSE)
```


# Appendix {-}
## Session Info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```