---
title: Annotating Neutrophil eQTL with Blueprint Data
author: Peter Humburg
date: `r format(Sys.time(), "%a %d %b %Y")`
---

```{r setup, include=FALSE}
library(GenomicRanges)
library(rtracklayer)
library(ggplot2)

opts_knit$set(root.dir = "/well/hill/Vivek/eQTL")
opts_knit$set(autodep=TRUE)
opts_knit$set(tidy=TRUE)
```

# Introduction
The aim of this analysis is to annotate eQTL obtained from the analysis
of neutrophils with the results of bisulfite sequencing and ChIP-seq
for several histone modifications generated by the Blueprint consortium.
BED files with ChIP-seq peaks and hyper- and hypo-methylated regions
all neutrophil samples were obtained form the Blueprint ftp server.

# Methylation data
## Data processing
The BED files provided by Blueprint do not conform to the BED specification.
This causes difficulties when trying to process them with BED file parsers.
To remedy this the first five columns of each file are extracted prior
to further analysis.

```{r reformatBED, engine="bash", echo=FALSE}
for file in $(ls blueprint/WGBS_Neut/*.bed)
do
	cut -f1-5 $file > ${file}5
done	
```
The re-formatted BED files are imported into R for further processing with 
Bioconductor. 

```{r loadWGBS, echo=FALSE, cache=TRUE}
methFiles <- dir("blueprint/WGBS_Neut", ".bed4", full.names=TRUE)
splitNames <- strsplit(basename(methFiles), ".")
sample <- sapply(splitNames, "[[", 1)
type <- sapply(splitNames, "[[", 2)
methDF <- data.frame(file=methFiles, sample=sample, type=type)
methGR <- do.call(GRangesList, 
		lapply(methFiles, import.bed, genome="hg19", asRangedData=FALSE))
```

## Defining regions of hyper- and hypo-methylation
Regions considered to be either hyper- or hypo-methylated are defined as follows.
A genomic region is considered to be hyper-methylated in neutrophils if it
meets the criteria for hyper-methylation used in the analysis of the Blueprint consortium
(average methylation of >0.75 and all CpGs in the regions have methylation >0.5)
in at least two neutrophil samplesand the region isn't hypo-methylated in any of 
the neutrophil samples. Hypo-methylated regions are defined accordingly.

```{r defineRegions, echo=FALSE, cache=TRUE}
allHypo <- do.call(union, methGR[methDF$type == "hypo_meth"])
allHyper <- do.call(union, methGR[methDF$type == "hyper_meth"])
hypo <- lapply(methGR[methDF$type == "hypo_meth"], setdiff, allHyper)
hyper <- lapply(methGR[methDF$type == "hyper_meth"], setdiff, allHypo)
```

```{r regionLength, echo=FALSE, fig.cap="Length distribution of hyper- and hypo-methylated regions"}
regLen <- list()
regLen$hypo <- data.frame(length=unlist(lapply(hypo, width), type="hypo methylated")
regLen$hyper <- data.frame(length=unlist(lapply(hyper, width), type="hyper methylated")
regLen <- rbind(regLen$hypo, regLen$hyper)
ggplot(regLen, aes(x=length)) + geom_histogram() + facet( ~ type) + theme_bw()
```

```{r overlapRegions, echo=FALSE, cache=TRUE}
hypo <- Reduce(subsetByOverlaps, hypo)
hyper <- Reduce(subsetByOverlaps, hyper)
```

# Appendix {-}
## Session Info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```